net: Add NAT-PMP port forwarding support

hebasto opened the pull request:
Close #11902
This PR is an alternative to:
- #12288
- #15717

To compile with NAT-PMP support on Ubuntu [`libnatpmp-dev`](https://packages.ubuntu.com/source/bionic/libnatpmp) should be available.

Log excerpt:
```
2020-02-05T20:12:28Z [mapport] NAT-PMP: public address = 95.164.65.194
2020-02-05T20:12:28Z [mapport] AddLocal(95.164.65.194:18333,3)
2020-02-05T20:12:28Z [mapport] NAT-PMP: port mapping successful.
```

See: [`libnatpmp`](https://miniupnp.tuxfamily.org/libnatpmp.html)

---

Some follow-ups are out of this PR's scope:
- mention NAT-PMP library in the version message
- ~integrate NAT-PMP into the GUI~ (already [added](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-589405068))

laanwj commented:
Thanks for picking this up again. I hope we manage to do this this time.

hebasto committed:
refactor: Move port mapping code to its own module

This commit does not change behavior.

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,221 @@
+    closenatpmp(&natpmp);
+}
+
+#else // USE_UPNP
```
This will break building with both NAT-PMP and UPnP support...

hebasto replied:
> This will break building with both NAT-PMP and UPnP support...

Yes, I'm aware of that. This PR in its current state is intended for NAT-PMP functionality tests and concept (N)ACKs.

luke-jr commented:
Concept ACK

hebasto commented:
It is ready for review now.

hebasto commented:
Updated 4ff5349a3172d02fceb976bf4989996c88f86680 -> cf94431e8f4bfd6636998a17e2355feb2c5c3d58 ([pr18077.03](https://github.com/hebasto/bitcoin/commits/pr18077.03) -> [pr18077.04](https://github.com/hebasto/bitcoin/commits/pr18077.04), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.03..pr18077.04)):

added `libnatpmp` package to macOS 10.14 native Travis build.

hebasto commented:
Updated cf94431e8f4bfd6636998a17e2355feb2c5c3d58 -> 88248a44241c8e3922a80729ce4a9a178ef2dbea ([pr18077.04](https://github.com/hebasto/bitcoin/commits/pr18077.04) -> [pr18077.05](https://github.com/hebasto/bitcoin/commits/pr18077.05), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.04..pr18077.05)):

added:
- changes to docs
- release notes

hebasto commented:
Rebased after #17398 and #18081 have been merged.

dongcarl commented:
Concept ACK, will test on my home router.

Sjors commented:
Concept ACK. Lightly tested configure / build on macOS.

Unfortunately I'm behind a IPv6 DS-Lite router, I can't test the IPv4 port forward. Perhaps `-7` could be replaced by something more readable.

```
2020-02-11T17:48:27Z natpmp thread start
2020-02-11T17:48:27Z Bound to [::]:8333
2020-02-11T17:48:27Z Bound to 0.0.0.0:8333
2020-02-11T17:48:27Z init message: Loading P2P addresses...
2020-02-11T17:48:27Z NAT-PMP: readnatpmpresponseorretry() for public address failed with -7 error.
2020-02-11T17:48:27Z NAT-PMP: readnatpmpresponseorretry() for port mapping failed with -7 error. 
```

Two suggesitons for followups, or this PR:
* add checkbox in QT Settings (above UPNP)
* add support for IPv6 pinhole if the library supports it, see #17012

hebasto commented:
@Sjors 

Thank you for your review and testing.

> I looked up the error code, and it's what I would expect given my setup:
> 
> ```
> /* NATPMP_ERR_NOGATEWAYSUPPORT : the gateway does not support NAT-PMP */
> #define NATPMP_ERR_NOGATEWAYSUPPORT (-7)
> ```

Agree.
But in this initial PR I intentionally do not provide user-friendly error messages in order to keep `ThreadNatpmp()` dense and tight for easier reviewing ;)

In followups every library call (`initnatpmp()`, `sendpublicaddressrequest()` and `sendnewportmappingrequest()`) with the following `readnatpmpresponseorretry()` loops could be refactored out to separate functions with user-friendly error messages.

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,236 @@
+        }
+        case MapPort::UPNP: {
+#ifdef USE_UPNP
+            g_mapport_thread = std::thread((std::bind(&TraceThread<void (*)()>, "upnp", &ThreadUpnp)));
```
Won't this overwrite the `g_mapport_thread` set above if both are enabled?

hebasto replied:
I believe that the both port mapping protocols shouldn't be enabled simultaneously:
https://github.com/bitcoin/bitcoin/blob/c7e90ca40ad9b7c66f3429b792435e9279ceca6b/src/init.cpp#L1781-L1786

luke-jr replied:
It should be possible to enable both.

Sjors replied:
If both are enabled, would you try to open de port with NAT_PMP first and if that fails UPNP?

hebasto replied:
> It should be possible to enable both.

Do you mean that a router could do port mapping using both NAT-PMP and UPnP simultaneously? What is reason for that?

Sjors replied:
I wouldn't use them simultaneously, but one fails it makes sense to try the other. Though I believe our long plan is to remove UPnP, if NAT-PMP does the job.

hebasto replied:
> I wouldn't use them simultaneously, but one fails it makes sense to try the other.

Good suggestion for a followup.

> Though I believe our long plan is to remove UPnP, if NAT-PMP does the job.

Not all routers with stock software have NAT-PMP port mapping feature.

luke-jr replied:
Users who don't know what their router supports should be able to just flip both on. And laptops might very well migrate between different routers using each protocol.

Agree with @hebasto that NAT-PMP can't be a complete substitute for UPnP since routers might not support both.

hebasto replied:
Addressed in the latest push.

hebasto committed:
refactor: Replace magic number with named constant

hebasto commented:
Updated c7e90ca40ad9b7c66f3429b792435e9279ceca6b -> dc228d6db8d2e9406d6d6e0eb04e37940b1ffb82 ([pr18077.06](https://github.com/hebasto/bitcoin/commits/pr18077.06) -> [pr18077.07](https://github.com/hebasto/bitcoin/commits/pr18077.07), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.06..pr18077.07)):

Changes:
- added user-friendly error messages for `NATPMP_ERR_NOGATEWAYSUPPORT` error
- enable both NAT-PMP and UPnP if `-natpmp` and `-upnp` options are supplied

---

@Sjors 
> Perhaps `-7` could be replaced by something more readable.

Done.

> If both are enabled, would you try to open de port with NAT_PMP first and if that fails UPNP?
> I wouldn't use them simultaneously, but one fails it makes sense to try the other.

Done.

luke-jr commented on src/init.cpp:
```diff
@@ -438,15 +439,24 @@ void SetupServerArgs()
     gArgs.AddArg("-upnp", "Use UPnP to map the listening port (default: 1 when listening and no -proxy)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
 #else
-    gArgs.AddArg("-upnp", strprintf("Use UPnP to map the listening port (default: %u)", 0), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-upnp", "Use UPnP to map the listening port (default: 0)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Prefer to leave this alone (and use `%s` in the new options)

hebasto replied:
Agree. It will be fixed.

hebasto replied:
Done.

luke-jr commented:
The logic here still seems broken. Again, consider that users may migrate between different routers each with NAT-PMP or UPnP support. So there should be a common thread main that *at each refresh* (which ideally should trigger when network availability changes, but that can be another PR) checks for which protocol to use and does the right thing.

hebasto commented:
@luke-jr 
> The logic here still seems broken. Again, consider that users may migrate between different routers each with NAT-PMP or UPnP support. So there should be a common thread main that _at each refresh_ (which ideally should trigger when network availability changes, but that can be another PR) checks for which protocol to use and does the right thing.

`ThreadAnyAvailable()` does almost the same:
if a used protocol becomes unavailable, `ThreadAnyAvailable()` switches to another one, no?

luke-jr commented:
Yes, re-reading it, you're right, it does. I saw the loops in the per-protocol functions and missed the outer one.

hebasto commented:
Updated dc228d6db8d2e9406d6d6e0eb04e37940b1ffb82 -> 2d5d98ce0aa07401c6771731f4e27634629f197c ([pr18077.07](https://github.com/hebasto/bitcoin/commits/pr18077.07) -> [pr18077.08](https://github.com/hebasto/bitcoin/commits/pr18077.08), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.07..pr18077.08)):

Changes:
- addressed @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r379960033)

luke-jr commented on src/init.cpp:
```diff
@@ -1764,9 +1778,17 @@ bool AppInitMain(NodeContext& node)
+        const bool natpmp = gArgs.GetBoolArg("-natpmp", DEFAULT_NATPMP);
+        const bool upnp = gArgs.GetBoolArg("-upnp", DEFAULT_UPNP);
+        if (natpmp && upnp) {
+            StartMapPort(MapPort::ANY_AVAILABLE);
```
Suggest just making this `StartMapPort(bool use_upnp, bool use_natpmp)`

luke-jr replied:
Or better yet: Just check gArgs inside StartMapPort...

hebasto replied:
Reworked in the latest push.

luke-jr commented on src/qt/optionsmodel.cpp:
```diff
@@ -350,9 +362,13 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             break;
+        case MapPortNatpmp: // core option - can be changed on-the-fly
+            settings.setValue("fUseNatpmp", value.toBool());
+            m_node.mapPort(value.toBool(), MapPort::NAT_PMP);
```
This logic won't work if the thread is already running...

hebasto replied:
Going to fix it.

hebasto replied:
Fixed in the latest push.

luke-jr commented on src/init.cpp:
```diff
@@ -438,6 +439,15 @@ void SetupServerArgs()
+#if USE_NATPMP
+    gArgs.AddArg("-natpmp", "Use NAT-PMP to map the listening port (default: 1 when listening and no -proxy)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+#else
+    gArgs.AddArg("-natpmp", strprintf("Use NAT-PMP to map the listening port (default: %u)", 0), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
`%s` > `%u`

hebasto replied:
> `%s` > `%u`

What is the rationale? We already use the `%u` specifier for unsigned integers in `tfm::format()`.

luke-jr replied:
`%s` doesn't care what type the value is.

You can use the same format string above, with a value of `"1 when listening and no -proxy"`

hebasto replied:
Done in the latest push.

hebasto committed:
net: Keep trying to use UPnP when -upnp=1

hebasto commented:
Updated 2d5d98ce0aa07401c6771731f4e27634629f197c -> ccb4a7ea18c38914f084b7c070303d020adad326 ([pr18077.08](https://github.com/hebasto/bitcoin/commits/pr18077.08) -> [pr18077.09](https://github.com/hebasto/bitcoin/commits/pr18077.09), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.08..pr18077.09)):

- fixed a [bug](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r380352965) pointed by @luke-jr

dongcarl commented:
Hi all, did some testing. What I did:

1. Installed natpmpd on my OpenBSD gateway router
2. Started natpmpd and add relevant pf.conf line
3. Configure and compile `bitcoind` with `--without-miniupnpc --disable-bench --disable-wallet --without-gui --with-natpmp`
4. Start `bitcoind` with `-natpmp`
5. Saw the following lines:
```
2020-02-18T22:26:26Z NAT-PMP: Public address = 108.21.84.253
2020-02-18T22:26:26Z AddLocal(108.21.84.253:8333,3)
2020-02-18T22:26:26Z NAT-PMP: Port mapping failed.
2020-02-18T22:26:26Z NAT-PMP: Port mapping removed successfully.
```
6. Checked my `natpmpd` anchored `pfctl` rules:
```
$ doas pfctl -a natpmpd -s rules
pass in quick on rdomain 0 inet proto tcp from any to 108.21.84.253 port = 62774 flags any rdr-to 192.168.0.5 port 8333
```

I believe this means that natpmpd successfully allocated a port 62774 to `bitcoind`, but since that port does not equal our private port, this check failed: https://github.com/bitcoin/bitcoin/pull/18077/files#diff-161ad6a962291e0bbafeb83ddc8a977cR109. I believe this is because `natpmpd` ignores our "Suggested External Port", which is perfectly valid, and we should probably account for that possibility.

luke-jr commented:
If NAT-PMP allows ignoring the suggested port, then we probably should handle that by advertising the correct port for the external IP in addrman.

hebasto commented:
@dongcarl 
Thank you for testing this PR with [`natpmpd`](https://github.com/bodgit/natpmpd) on OpenBSD (I've tested with [`miniupnpd`](https://github.com/miniupnp/miniupnp/tree/master/miniupnpd) on OpenWrt).

> I believe this is because `natpmpd` ignores our "Suggested External Port", which is perfectly valid, and we should probably account for that possibility.

Correct. From `natpmpd` [docs](http://bodgitandscarper.co.uk/natpmpd/):
> The external TCP or UDP port assigned to the client is _always_ randomised rather than giving the first client the port it actually requested and then trying to work out what to do for additional clients that want the same external port.

(highlighted in italics by me).

OTOH, [RFC 6886](https://tools.ietf.org/html/rfc6886) states:
> If the client would prefer to have a high-numbered "anonymous" external port assigned, then it should set the Suggested External Port to zero, which indicates to the gateway that it should allocate a high-numbered port of its choosing.  If the client would prefer instead to have the mapped external port be the same as its local internal port if possible (e.g., a web server listening on port 80 that would ideally like to have external port 80), then it should set the Suggested External Port to the desired value.  However, the gateway is _not_ obliged to assign the port suggested, and may choose not to, either for policy reasons (e.g., port 80 is reserved and clients may not request it) or because that port has already been assigned to some other client.

(highlighted in italics by me).

@luke-jr 
> If NAT-PMP allows ignoring the suggested port, then we probably should handle that by advertising the correct port for the external IP in addrman.

Sure, we should.

hebasto commented:
Updated ccb4a7ea18c38914f084b7c070303d020adad326 -> 03e2733be1772ef69caf34e84c7b400f64a8ea9f ([pr18077.09](https://github.com/hebasto/bitcoin/commits/pr18077.09) -> [pr18077.10](https://github.com/hebasto/bitcoin/commits/pr18077.10), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.09..pr18077.10)):

- aligned with RFC 6886

@dongcarl @luke-jr 
Thank you for your reviews. All your comments have been addressed.

luke-jr commented on src/qt/optionsmodel.cpp:
```diff
@@ -344,15 +356,19 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             break;
+        case MapPortNatpmp: // core option - can be changed on-the-fly
+            settings.setValue("fUseNatpmp", value.toBool());
+            m_node.mapPort(value.toBool(), settings.value("fUseUPnP").toBool());
```
This logic breaks the "command line overrides GUI options" stuff above (see `addOverriddenOption`)...

Not really sure what the best solution here is.

hebasto replied:
The separate issue #18184 has been submitted.

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,320 @@
+        if (g_mapport_target_proto == MapPortProto::NONE) {
+            return;
+        }
+    } else if (g_mapport_target_proto & g_mapport_current_proto) {
```
This won't work if we used to have both protocols enabled, are currently using UPnP, but want to disable UPnP.

hebasto replied:
Sure it works.
The log excerpt for the mentioned case:
```
2020-02-20T23:26:30Z [mapport] UPNP_DeletePortMapping() returned: 0
2020-02-20T23:26:30Z [mapport] AddLocal(95.164.65.194:18333,3)
2020-02-20T23:26:30Z [mapport] NAT-PMP: Port mapping successful. External address = 95.164.65.194:18333

```

luke-jr replied:
Why would it? You'd return early here without making the change...

hebasto replied:
If both protocols were enabled, `g_mapport_target_proto != MapPortProto::NONE`

luke-jr replied:
The conditional here is `(g_mapport_target_proto & g_mapport_current_proto)`

hebasto replied:
```C++
void StartMapPort(bool use_natpmp, bool use_upnp)
{
    // Both protocols are enabled, so
    // g_mapport_target_proto == MapPortProto::NAT_PMP | MapPortProto::UPNP
    //
    // UPnP is currently using, so
    // g_mapport_current_proto == MapPortProto::UPNP
    //
    // A user disables UPnP then passed arguments are:
    // use_natpmp == true, use_upnp = false

    if (use_natpmp) {
        // Effectively noop here as MapPortProto::NAT_PMP is set already.
        g_mapport_target_proto |= MapPortProto::NAT_PMP;
    } else {
        g_mapport_target_proto &= ~MapPortProto::NAT_PMP;
    }

    if (use_upnp) {
        g_mapport_target_proto |= MapPortProto::UPNP;
    } else {
        // Unset MapPortProto::UPNP bit.
        g_mapport_target_proto &= ~MapPortProto::UPNP;
    }

    // At this point variables are set to the following values:
    // g_mapport_target_proto == MapPortProto::NAT_PMP
    // g_mapport_current_proto == MapPortProto::UPNP
    // ...
    if (g_mapport_current_proto == MapPortProto::NONE) {
        if (g_mapport_target_proto == MapPortProto::NONE) {
            return;
        }
    // ... and the next condition evaluates to false.
    } else if (g_mapport_target_proto & g_mapport_current_proto) {
        return;
    }

    // Making the changes...
    ...
}
```

And logs confirm this logic.

luke-jr replied:
Got it. I was confusing it with (target & new_target)

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,320 @@
+    }
+
+    if (g_mapport_thread.joinable()) {
+        g_mapport_interrupt();
```
Won't t his cause the thread to shut down?

hebasto replied:
Fixed in the latest [push](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-589405068).

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,320 @@
+        bool external_ip_discovered = false;
+        const uint16_t private_port = GetListenPort();
+        do {
+            ret = NatpmpMapping(&natpmp, external_ipv4_addr, private_port, external_ip_discovered);
```
I don't see the logic here to handle being assigned a different port number.

hebasto replied:
The logic to handle being assigned a different external port number by the gateway uses the `g_mapport_external_port` variable.

luke-jr commented:
I suggest putting the variable renames in a separate commit for easier review.

Also, since NAT-PMP apparently maps random ports often, I think we should prefer UPnP...

hebasto commented:
@dongcarl 
Thank you for testing this PR with [`natpmpd`](https://github.com/bodgit/natpmpd) on OpenBSD (I've tested with [`miniupnpd`](https://github.com/miniupnp/miniupnp/tree/master/miniupnpd) on OpenWrt).

> I believe this is because `natpmpd` ignores our "Suggested External Port", which is perfectly valid, and we should probably account for that possibility.

Correct. From `natpmpd` [docs](http://bodgitandscarper.co.uk/natpmpd/):
> The external TCP or UDP port assigned to the client is _always_ randomised rather than giving the first client the port it actually requested and then trying to work out what to do for additional clients that want the same external port.

(highlighted in italics by me).

OTOH, [RFC 6886](https://tools.ietf.org/html/rfc6886) states:
> If the client would prefer to have a high-numbered "anonymous" external port assigned, then it should set the Suggested External Port to zero, which indicates to the gateway that it should allocate a high-numbered port of its choosing.  If the client would prefer instead to have the mapped external port be the same as its local internal port if possible (e.g., a web server listening on port 80 that would ideally like to have external port 80), then it should set the Suggested External Port to the desired value.  However, the gateway is _not_ obliged to assign the port suggested, and may choose not to, either for policy reasons (e.g., port 80 is reserved and clients may not request it) or because that port has already been assigned to some other client.

(highlighted in italics by me).

@luke-jr 
> If NAT-PMP allows ignoring the suggested port, then we probably should handle that by advertising the correct port for the external IP in addrman.

Sure, we should.

hebasto commented:
@luke-jr 
> I suggest putting the variable renames in a separate commit for easier review.

There is a dedicated commit "refactor: Rename UPnP stuff" (a2de373a96dc1b9f5374eb70d605dcaf49f9b244).
What variable rename did I miss to put it to?

hebasto commented:
Updated 03e2733be1772ef69caf34e84c7b400f64a8ea9f -> dd3229e7af8563c69d26955c4207f73ae1d3bc94 ([pr18077.10](https://github.com/hebasto/bitcoin/commits/pr18077.10) -> [pr18077.11](https://github.com/hebasto/bitcoin/commits/pr18077.11), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.10..pr18077.11)):

- fixed CI issues
- the NAT-PMP checkbox added to the GUI (this simplifies testing a lot)
- fixed a [bug](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r381558152) pointed by @luke-jr

hebasto commented:
@luke-jr 
> Also, since NAT-PMP apparently maps random ports often, I think we should prefer UPnP...

Does this behavior of NAT-PMP gateways hurt the entire p2p network?

OTOH, NAT-PMP seems more secure than UPnP (for example, one could compare numbers of CVEs for each protocol).

luke-jr commented:
>Does this behavior of NAT-PMP gateways hurt the entire p2p network?

DNS seeds can't provide port numbers, so using anything other than 8333 reduces the list of nodes we can advertise. Currently, mine (which is IIRC more restrictive than others) has 3090 eligible nodes, so it's not a big deal, but it'd be nicer to have a wider selection available than a narrower selection.

> OTOH, NAT-PMP seems more secure than UPnP (for example, one could compare numbers of CVEs for each protocol).

The CVEs AFAIK are for the implementations, not the protocol.

In any case, when we are using either, I think we're reduced to the weakest no matter which we prefer (a hostile router could block NAT-PMP to force fallback to UPnP).

luke-jr commented:
Re NAT-PMP preference: Also, with the port number changing every startup, it will be hard for peers to reconnect or meaningfully pass on addr messages for you...

I think cd7a0f689c5 should get split up. It's too hard to follow what it's doing.

hebasto commented:
> Re NAT-PMP preference: Also, with the port number changing every startup, it will be hard for peers to reconnect or meaningfully pass on addr messages for you...

Only [`natpmpd`](https://github.com/bodgit/natpmpd) on OpenBSD will change external port number on every startup.
[`miniupnpd`](https://github.com/miniupnp/miniupnp/tree/master/miniupnpd) assigns suggested one, i.e. 8333.

hebasto commented on src/mapport.cpp:
```diff
@@ -0,0 +1,322 @@
+            return;
+        }
+
+    } while (ok || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD));
```
If the protocol switch is initiated by a user, the `g_mapport_interrupt()` call will end the internal protocol-specific loop in the `ThreadNatpmp()` or in the `ThreadUpnp()`, `ok == true` and `g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD)` will not call at all.

If the protocol switch is initiated by a protocol failure, `ok == false` and `g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD)` will behave as intended.

Setting `g_mapport_target_proto` variable to `MapPortProto::NONE` also causes shutting off port mapping entirely.

luke-jr commented:
Yes, but UPnP is reliably using the same/requested port if it works.

hebasto committed:
net: Add flags for port mapping protocols

hebasto committed:
scripted-diff: Rename UPnP stuff

-BEGIN VERIFY SCRIPT-
sed -i 's/g_upnp_interrupt/g_mapport_interrupt/' src/mapport.cpp
sed -i 's/if(g_upnp_thread/if (g_mapport_thread/' src/mapport.cpp
sed -i 's/g_upnp_thread/g_mapport_thread/' src/mapport.cpp
sed -i 's/LOCAL_UPNP/LOCAL_MAPPED/' src/mapport.cpp
sed -i 's/\bupnp\b/mapport/' src/mapport.cpp
sed -i 's/LOCAL_UPNP,  /LOCAL_MAPPED,/' src/net.h
-END VERIFY SCRIPT-

hebasto committed:
gui: Apply port mapping changes on dialog exit

This commit does not change behavior. It is a prerequisite for NAT-PMP
support adding.

hebasto committed:
net: Add libnatpmp support

hebasto committed:
net: Add NAT-PMP to port mapping loop

hebasto committed:
net: Add -natpmp command line option

hebasto committed:
gui: Add NAT-PMP network option

hebasto committed:
ci: Add libnatpmp-dev package to some builds

hebasto committed:
doc: Add libnatpmp stuff

hebasto committed:
doc: Add release notes

hebasto commented:
Updated dd3229e7af8563c69d26955c4207f73ae1d3bc94 -> d9ecda8fa0ebce5b6da972691b19f4425164ae43 ([pr18077.11](https://github.com/hebasto/bitcoin/commits/pr18077.11) -> [pr18077.12](https://github.com/hebasto/bitcoin/commits/pr18077.12), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.11..pr18077.12)):

- UPnP is prioritized (as @luke-jr suggested)
- a note about possible random external port number is added to the NAT-PMP tooltip
- commits are split and reordered to make review easier, I hope

hebasto commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+        do {
+            ret = NatpmpMapping(&natpmp, external_ipv4_addr, private_port, external_ip_discovered);
+        } while (ret && g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+        g_mapport_interrupt.reset();
```
@luke-jr 
> This should be reset outside of the protocol-specific functions, or else we could end up with a race.

Looking into the `CThreadInterrupt` class implementation I couldn't see possibilities for a race.

What did I miss?

hebasto commented on src/mapport.cpp:
```diff
@@ -115,14 +117,39 @@ static void ThreadMapPort()
 }
 
-void StartMapPort()
+void StartThread()
```
> This name is way too vague...

Yeah... I'm not good at choosing English names, and will appreciate your suggestion.

luke-jr replied:
StartThreadMapPort maybe

hebasto replied:
Done.

hebasto commented on src/mapport.h:
```diff
@@ -12,7 +12,12 @@ static const bool DEFAULT_UPNP = USE_UPNP;
 #endif
 
-void StartMapPort();
+enum MapPortProto : unsigned int {
```
It will require to define bitwise operators which are used here:
https://github.com/bitcoin/bitcoin/blob/d9ecda8fa0ebce5b6da972691b19f4425164ae43/src/mapport.cpp#L297-L304

Maybe in follow-up?

luke-jr replied:
I wish there was a C++ standard QFlags...

hebasto commented on src/qt/optionsdialog.cpp:
```diff
@@ -50,6 +51,10 @@ OptionsDialog::OptionsDialog(QWidget *parent, bool enableWallet) :
 #endif
+    connect(this, &QDialog::accepted, [this](){
+        QSettings settings;
+        model->node().mapPort(settings.value("fUseUPnP").toBool());
```
> Are we sure the settings are changed _before_ the signal? Or could it be in parallel and therefore a race?

Yes, we are. Please consider the following slot:
https://github.com/bitcoin/bitcoin/blob/d9ecda8fa0ebce5b6da972691b19f4425164ae43/src/qt/optionsdialog.cpp#L284-L289

All changes are submitted to the model in the `mapper->submit()` call _before_ the `QDialog::accepted()` signal is emitted when the dialog has been accepted by calling `accept()`.

Then the following connection:
https://github.com/bitcoin/bitcoin/blob/d9ecda8fa0ebce5b6da972691b19f4425164ae43/src/qt/optionsdialog.cpp#L57-L60
does its work.

hebasto commented:
Updated d9ecda8fa0ebce5b6da972691b19f4425164ae43 -> 405e85c61a6690ff2f7159c51715fdbafcbf2b1e ([pr18077.12](https://github.com/hebasto/bitcoin/commits/pr18077.12) -> [pr18077.13](https://github.com/hebasto/bitcoin/commits/pr18077.13), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.12..pr18077.13)):

- `StartThread()` renamed to `StartThreadMapPort()` (as @luke-jr [suggested)](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r383498346)
- conditional operator argument types are the same now (https://github.com/bitcoin/bitcoin/pull/18077#discussion_r383034794)
- renamed `MapPortProto` enum to `MapPortProtoFlag` and `g_mapport_target_proto` to `g_mapport_enabled_protos` for better semantics

hebasto commented on src/init.cpp:
```diff
@@ -447,7 +447,12 @@ void SetupServerArgs()
-#endif
+#endif // #ifdef USE_UPNP
+#ifdef USE_NATPMP
+    gArgs.AddArg("-natpmp", strprintf("Use NAT-PMP to map the listening port (default: %s)", DEFAULT_NATPMP ? "1 when listening and no -proxy" : 0), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Fixed.

hebasto commented on src/init.cpp:
```diff
@@ -447,7 +447,12 @@ void SetupServerArgs()
-#endif
+#endif // #ifdef USE_UPNP
+#ifdef USE_NATPMP
+    gArgs.AddArg("-natpmp", strprintf("Use NAT-PMP to map the listening port (default: %s)", DEFAULT_NATPMP ? "1 when listening and no -proxy" : 0), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Fixed.

sipa commented on src/mapport.cpp:
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2020 The Bitcoin Core developers
```
In commit "refactor: Move port mapping code to own files"

Nit: when you're copying/moving code, retain the copyright years of the original code.

hebasto replied:
The UPnP support code has been added in 2011 (see #133).
Is it correct to retain pre-2011 copyright years?

Could it be "2011-2020" in new files?

hebasto replied:
See the latest [push](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-591285675).

sipa commented on src/mapport.cpp:
```diff
@@ -35,9 +35,11 @@ static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed"
+static constexpr auto PORT_MAPPING_RETRY_PERIOD = std::chrono::minutes(5);
 
-static void ThreadMapPort()
+static bool ThreadUpnp()
```
In commit "net: Keep trying to use UPnP when -upnp=1"

This is confusingly named when it isn't actually a thread on its own anymore.

hebasto replied:
[Renamed](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-591285675).

sipa reviewed (commented):
Concept ACK, left a few code review nits. I haven't reviewed the actual natpmp interaction code.

hebasto commented:
Updated 405e85c61a6690ff2f7159c51715fdbafcbf2b1e -> 4e7b9f03b8a3dd02bc9831e77905a8022770c846 ([pr18077.13](https://github.com/hebasto/bitcoin/commits/pr18077.13) -> [pr18077.14](https://github.com/hebasto/bitcoin/commits/pr18077.14), [compare](https://github.com/hebasto/bitcoin/compare/pr18077.13..pr18077.14)):

- corrected copyright years as @sipa [suggested](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r384206803)
- both `ThreadUpnp()` and `ThreadNatpmp()` renamed to `ProcessUpnp()` and `ProcessNatpmp()` respectively (see @sipa's [comment](https://github.com/bitcoin/bitcoin/pull/18077#discussion_r384207039))
- removed extra parens (thanks to @Empact; see https://github.com/bitcoin/bitcoin/pull/12381#discussion_r167062893)

hebasto commented:
Rebased after #17399 has been merged.

hebasto commented:
Rebased due to the conflict with merged #18264.

hebasto commented:
Rebased 4be5c3c72d98b616367d1aa3de22702e6f9f81ee -> 1bf61f0557637b5bedd670a4fba7bb11bf42ac6a due to the conflicts with #18134 and #18438.

hebasto commented:
Rebased 1bf61f0557637b5bedd670a4fba7bb11bf42ac6a -> 607144bab1aecb8b970e17aeb145d62175d58c4c ([pr18077.17](https://github.com/hebasto/bitcoin/commits/pr18077.17) -> [pr18077.18](https://github.com/hebasto/bitcoin/commits/pr18077.18)) due to the conflicts with #18514 and #16367.

hebasto commented:
Rebased 607144bab1aecb8b970e17aeb145d62175d58c4c -> a388fae1dca9b4983bc95d50d07503aae527f29e ([pr18077.18](https://github.com/hebasto/bitcoin/commits/pr18077.18) -> [pr18077.19](https://github.com/hebasto/bitcoin/commits/pr18077.19)) due to the conflict with #18588.

hebasto commented:
Rebased a388fae1dca9b4983bc95d50d07503aae527f29e -> 95e943cb1855dca3ed5fb4fd98b357515d1a3646 ([pr18077.19](https://github.com/hebasto/bitcoin/commits/pr18077.19) -> [pr18077.20](https://github.com/hebasto/bitcoin/commits/pr18077.20)) due to the conflict with #18571.

hebasto commented:
Rebased 95e943cb1855dca3ed5fb4fd98b357515d1a3646 -> f4a4643712eca819a5b7ed0a143216e06125d1a9 ([pr18077.20](https://github.com/hebasto/bitcoin/commits/pr18077.20) -> [pr18077.21](https://github.com/hebasto/bitcoin/commits/pr18077.21)) due to the conflicts with #18676 and #18705.

tryphe commented:
tACK f4a4643712eca819a5b7ed0a143216e06125d1a9 on Debian 9

```
2020-05-17T04:14:52Z NAT-PMP: Port mapping successful. External address = [redacted]:8333
2020-05-17T04:22:05Z NAT-PMP: Port mapping removed successfully.
```

The port is open without any other forwarding rules.

hebasto commented:
Rebased f4a4643712eca819a5b7ed0a143216e06125d1a9 -> 2f7a01708482ad9f8fede1d7b7a9d59901370aa1 ([pr18077.21](https://github.com/hebasto/bitcoin/commits/pr18077.21) -> [pr18077.22](https://github.com/hebasto/bitcoin/commits/pr18077.22)) due to the conflict with #19008.

hebasto commented:
Rebased 2f7a01708482ad9f8fede1d7b7a9d59901370aa1 -> f0cf6523c9fb4c3bb0bc64af72cdcadfa7747748 ([pr18077.22](https://github.com/hebasto/bitcoin/commits/pr18077.22) -> [pr18077.23](https://github.com/hebasto/bitcoin/commits/pr18077.23)) due to the conflict with #18677.

hebasto commented:
Rebased f0cf6523c9fb4c3bb0bc64af72cdcadfa7747748 -> 1b006d618d02eac38bd434937d6d5c42084992bf ([pr18077.23](https://github.com/hebasto/bitcoin/commits/pr18077.23) -> [pr18077.24](https://github.com/hebasto/bitcoin/commits/pr18077.24)) due to the conflict with #19014.

hebasto commented:
Rebased 1b006d618d02eac38bd434937d6d5c42084992bf -> 0692bed243adb8d4f3aa8192e9d1952d1de4a2ad ([pr18077.24](https://github.com/hebasto/bitcoin/commits/pr18077.24) -> [pr18077.25](https://github.com/hebasto/bitcoin/commits/pr18077.25)) due to the conflict with #19041.

hebasto commented:
Rebased 0692bed243adb8d4f3aa8192e9d1952d1de4a2ad -> 5a29078b373f5ffecb452317a1bd6fb165b08b42 ([pr18077.25](https://github.com/hebasto/bitcoin/commits/pr18077.25) -> [pr18077.26](https://github.com/hebasto/bitcoin/commits/pr18077.26)) due to the conflict with #19180.

hebasto commented:
Rebased 5a29078b373f5ffecb452317a1bd6fb165b08b42 -> 570c1060058d7f52286107df217b66d28d0b1f91 ([pr18077.26](https://github.com/hebasto/bitcoin/commits/pr18077.26) -> [pr18077.27](https://github.com/hebasto/bitcoin/commits/pr18077.27)) due to the conflict with #19267.

hebasto commented:
Rebased 570c1060058d7f52286107df217b66d28d0b1f91 -> a660ecc26e691ed9b931d6fe27c26899ef66be64 ([pr18077.27](https://github.com/hebasto/bitcoin/commits/pr18077.27) -> [pr18077.28](https://github.com/hebasto/bitcoin/commits/pr18077.28)) due to the conflict with #19331.

hebasto commented:
Rebased a660ecc26e691ed9b931d6fe27c26899ef66be64 -> d4a143ae233fc5fd0cb3c08ce7c388434d256574 ([pr18077.28](https://github.com/hebasto/bitcoin/commits/pr18077.28) -> [pr18077.29](https://github.com/hebasto/bitcoin/commits/pr18077.29),  [diff](https://github.com/hebasto/bitcoin/compare/pr18077.28..pr18077.29)):

- fixed dependency `libnatpmp` build with the custom `CC` variable.

hebasto commented:
Rebased d4a143ae233fc5fd0cb3c08ce7c388434d256574 -> b61233d7ab4afe6cd7875b178e923fab59f0534a ([pr18077.29](https://github.com/hebasto/bitcoin/commits/pr18077.29) -> [pr18077.30](https://github.com/hebasto/bitcoin/commits/pr18077.30)) due to the conflicts with #19191 and #19314.

hebasto commented:
Rebased b61233d7ab4afe6cd7875b178e923fab59f0534a -> 1c4bfdb46d48fed6c3f43bbdd7973069cdd21baf ([pr18077.30](https://github.com/hebasto/bitcoin/commits/pr18077.30) -> [pr18077.31](https://github.com/hebasto/bitcoin/commits/pr18077.31)) due to the conflict with #19495.

hebasto commented:
Rebased 1c4bfdb46d48fed6c3f43bbdd7973069cdd21baf -> 155d87877c11a32abe1ed91799776691b5977a31 ([pr18077.31](https://github.com/hebasto/bitcoin/commits/pr18077.31) -> [pr18077.32](https://github.com/hebasto/bitcoin/commits/pr18077.32)) due to the conflict with #17919.

hebasto commented:
Rebased 155d87877c11a32abe1ed91799776691b5977a31 -> 88b1f0357c332bc67fa20450dda9f77fa449576e ([pr18077.32](https://github.com/hebasto/bitcoin/commits/pr18077.32) -> [pr18077.33](https://github.com/hebasto/bitcoin/commits/pr18077.33)) due to the conflict with #19561.

hebasto commented:
Rebased 88b1f0357c332bc67fa20450dda9f77fa449576e -> 225f512438f4391f6af255f851131358e3c466a2 ([pr18077.33](https://github.com/hebasto/bitcoin/commits/pr18077.33) -> [pr18077.34](https://github.com/hebasto/bitcoin/commits/pr18077.34)) due to the conflict with #19098.

adamjonas commented:
For those looking to review or revisit:

There are concept ACKs from: luke-jr, dongcarl, Sjors, Sipa
a tACK from tryphe
with no NACKs or blocking issues

dongcarl commented:
Tested 225f512438f4391f6af255f851131358e3c466a2, works:
```
2020-08-20T19:07:08Z NAT-PMP: Port mapping successful. External address = <redacted>
```

---

Did not read code yet

hebasto commented:
Rebased 225f512438f4391f6af255f851131358e3c466a2 -> 22cbf4eefaa7333a3e2bf94fbf4d3e7b4bf379cb ([pr18077.34](https://github.com/hebasto/bitcoin/commits/pr18077.34) -> [pr18077.35](https://github.com/hebasto/bitcoin/commits/pr18077.35)) due to the conflict with #19779.

hebasto commented:
Rebased 22cbf4eefaa7333a3e2bf94fbf4d3e7b4bf379cb -> 2b6bd32e915d13928577b038c18194ac4024fc10 ([pr18077.35](https://github.com/hebasto/bitcoin/commits/pr18077.35) -> [pr18077.36](https://github.com/hebasto/bitcoin/commits/pr18077.36)) due to the conflict with https://github.com/bitcoin-core/gui/pull/35.

hebasto commented:
Rebased 2b6bd32e915d13928577b038c18194ac4024fc10 -> 62601a5a158b8c3710fde7cac8aadaa71a72f110 ([pr18077.36](https://github.com/hebasto/bitcoin/commits/pr18077.36) -> [pr18077.37](https://github.com/hebasto/bitcoin/commits/pr18077.37)) due to the conflict with #19558.

jonatack commented:
Concept ACK

gmaxwell commented:
I performed basic functional testing against this branch (but not master with this applied due to rebase hell) against http://www.nongnu.org/natpmp/.

I don't really like the natpmp default being set by configure. I see the UPNP is currently handled that way, but I think that is an artefact of the fact that UPNP is disabled for implementation specific security reasons. Otherwise, no other bitcoin setting has its default set by configure.  That said, this could easily be something changed after getting the functionality merged or even in a subsequent release.

I found that it successfully mapped 8333 and released the mapping on shutdown.

I restricted the port range at the pmp-daemon and found that it successfully mapped a different port and correctly learned the other port number.

I killed the pmp-daemon while bitcoind was running and verified that it didn't hang up waiting for some unmap confirmation or anything like that.

I did not change my external IP to confirm that it learned about the IP change (as I couldn't do that without disrupting other people).

Looks to me like it works!

hebasto commented:
Rebased 62601a5a158b8c3710fde7cac8aadaa71a72f110 -> 44a697619d3fc3d9311c3d099cae04b21de071f3 ([pr18077.37](https://github.com/hebasto/bitcoin/commits/pr18077.37) -> [pr18077.38](https://github.com/hebasto/bitcoin/commits/pr18077.38)) due to the conflict with #19998.

gmaxwell commented:
ultra minor nit, ignore if you like:  I didn't like that the logs calls it "NAT-PMP" but the configuration option is natpmp, so I had to go read the code to figure out what I should be looking for. I'd prefer to see logs say "natpmp" for optimum greppability.

hebasto commented:
Rebased 44a697619d3fc3d9311c3d099cae04b21de071f3 -> 81a791d88d68e64968e52d8ff6ec20879bd5fad1 ([pr18077.38](https://github.com/hebasto/bitcoin/commits/pr18077.38) -> [pr18077.39](https://github.com/hebasto/bitcoin/commits/pr18077.39)) due to the conflict with #19077, and addressed @gmaxwell's [comment](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-709047605): 
> ultra minor nit, ignore if you like: I didn't like that the logs calls it "NAT-PMP" but the configuration option is natpmp, so I had to go read the code to figure out what I should be looking for. I'd prefer to see logs say "natpmp" for optimum greppability.

gmaxwell commented:
Still works. Looks good to me.

jamesob commented:
Concept ACK - reviewing this shortly

jamesob commented on src/mapport.cpp:
```diff
@@ -32,13 +41,117 @@ static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed"
+    natpmp_t natpmp;
+    struct in_addr external_ipv4_addr;
+    if (NatpmpInit(&natpmp) && NatpmpDiscover(&natpmp, external_ipv4_addr)) {
+        bool external_ip_discovered = false;
```
https://github.com/bitcoin/bitcoin/pull/18077/commits/f6bd717f3f44f2a167ce2e548df1132789d0ea98

Hm, I see we're explicitly setting this in `NetpmpMapping` but then it goes unused. No big deal, just curious.

hebasto replied:
As an in-out parameter, `external_ip_discovered` is used here https://github.com/bitcoin/bitcoin/blob/81a791d88d68e64968e52d8ff6ec20879bd5fad1/src/mapport.cpp#L136-L138 to prevent consecutive `AddLocal` calls.

jamesob replied:
Oh I see, thanks.

jamesob commented on src/init.cpp:
```diff
@@ -479,7 +479,12 @@ void SetupServerArgs(NodeContext& node)
-#endif
+#endif // #ifdef USE_UPNP
+#ifdef USE_NATPMP
+    argsman.AddArg("-natpmp", strprintf("Use NAT-PMP to map the listening port (default: %s)", DEFAULT_NATPMP ? "1 when listening and no -proxy" : "0"), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Nit: linebreak would be nice.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-714683741).

hebasto replied:
Reverted back as a linebreak breaks `test/lint/check-doc.py` linter: https://travis-ci.org/github/bitcoin/bitcoin/jobs/738110074

jamesob commented on src/qt/optionsdialog.cpp:
```diff
@@ -50,6 +50,9 @@ OptionsDialog::OptionsDialog(QWidget *parent, bool enableWallet) :
     ui->mapPortUpnp->setEnabled(false);
+#endif
+#ifndef USE_NATPMP
+    ui->mapPortNatpmp->setEnabled(false);
```
Nit: not s/false/USE_NATPMP ?

hebasto replied:
It won't compile if `USE_NATPMP` is undefined. Did I understand you correctly?

jamesob replied:
Oh my mistake; I read that as ifdef.

jamesob reviewed (approved):
Code review ACK 81a791d88d68e64968e52d8ff6ec20879bd5fad1

Really worthwhile change; all else equal the easier we can make offering inbound conn capacity the better.

I've reviewed and built locally, though my router doesn't support NAT-PMP out of the box. I'll see what I can do in the meantime in terms of more comprehensive testing.

hebasto commented on src/net.h:
```diff
@@ -33,11 +33,6 @@
 #include <condition_variable>
 
-#ifndef WIN32
-#include <arpa/inet.h>
```
Yes. I've submitted a dedicated #20221.

hebasto commented on src/mapport.cpp:
```diff
@@ -35,9 +35,11 @@ static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed"
 static CThreadInterrupt g_upnp_interrupt;
```
The _"net: Keep trying to use UPnP when -upnp=1"_ commit (5b6af780967c428ec38fa18ae68a193d2d036f57) is a prerequisite for the _"net: Add NAT-PMP to port mapping loop"_ commit (c329ae5b6834e8dae6d178bc84321c7dc6e3c6fc).
Both commits make possible to switch port forwarding method at runtime.

hebasto commented:
Rebased 81a791d88d68e64968e52d8ff6ec20879bd5fad1 -> cb4060686155c011cbece84fd551dfa000277d51 ([pr18077.39](https://github.com/hebasto/bitcoin/commits/pr18077.39) -> [pr18077.40](https://github.com/hebasto/bitcoin/commits/pr18077.40), [diff](https://github.com/hebasto/bitcoin/compare/pr18077.39..pr18077.40)).

Addressed @jamesob's comments:
-  https://github.com/bitcoin/bitcoin/pull/18077#discussion_r508829874:
> (Note on [1bf23f2](https://github.com/bitcoin/bitcoin/commit/1bf23f291d7c053538245f1a7f6f7699daf4a3b9) as a whole:) could be a scripteddiff AFAICT, but not a huge deal.

- https://github.com/bitcoin/bitcoin/pull/18077#discussion_r508839477:
> Nit: linebreak would be nice.

hebasto commented on src/mapport.cpp:
```diff
@@ -33,8 +33,8 @@ static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed"
 #ifdef USE_UPNP
-static CThreadInterrupt g_upnp_interrupt;
-static std::thread g_upnp_thread;
+static CThreadInterrupt g_mapport_interrupt;
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-714683741).

hebasto commented:
Rebased cb4060686155c011cbece84fd551dfa000277d51 -> bddfeade1e7740f04002f6d88e6e9840ed985bea ([pr18077.40](https://github.com/hebasto/bitcoin/commits/pr18077.40) -> [pr18077.41](https://github.com/hebasto/bitcoin/commits/pr18077.41), [diff](https://github.com/hebasto/bitcoin/compare/pr18077.40..pr18077.41)): https://github.com/bitcoin/bitcoin/pull/18077#discussion_r510714033

hebasto commented:
Rebased bddfeade1e7740f04002f6d88e6e9840ed985bea -> 5fd127243f452827dab3b4f4f47b6e96816fdfb8 ([pr18077.41](https://github.com/hebasto/bitcoin/commits/pr18077.41) -> [pr18077.42](https://github.com/hebasto/bitcoin/commits/pr18077.42)) due to the conflict with #19124.

jamesob reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/commit/5fd127243f452827dab3b4f4f47b6e96816fdfb8

Re-reviewed diffs after rebase. Built locally with NAT-PMP enabled. My gateway doesn't support NAT-PMP, but regular IBD works okay.

hebasto commented:
Rebased 5fd127243f452827dab3b4f4f47b6e96816fdfb8 -> 90d9547f135ef0ffc00dd790072c3ae307f053ad ([pr18077.42](https://github.com/hebasto/bitcoin/commits/pr18077.42) -> [pr18077.43](https://github.com/hebasto/bitcoin/commits/pr18077.43)) due to the conflict with #20156.

luke-jr commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    assert(g_mapport_thread.joinable());
+    assert(!g_mapport_interrupt);
```
Couldn't this be true if the user is impossibly fast?

hebasto replied:
I don't think so due to the `CThreadInterrupt` class implementation.

hebasto commented on doc/release-notes-18077.md:
```diff
@@ -0,0 +1,9 @@
+Command-line options
+--------------------
+
+-  The `-natpmp` option has been added to use NAT-PMP to map the listening port. If both `-upnp` and `-natpmp` options are provided, the former prevails.
```
Simultaneously?

hebasto commented:
Rebased 29fb5708ca04da9ca0c25c1b801887354eb714bc -> 3c26d53f538442363834178a45bc8b459b5b334d ([pr18077.44](https://github.com/hebasto/bitcoin/commits/pr18077.44) -> [pr18077.45](https://github.com/hebasto/bitcoin/commits/pr18077.45)) due to the conflicts with #20202, #20494 and the recent CI changes.

hebasto commented:
Rebased 3c26d53f538442363834178a45bc8b459b5b334d -> cc68169e72c5b28fe53466a86eeac9fd214cd0d1 ([pr18077.45](https://github.com/hebasto/bitcoin/commits/pr18077.45) -> [pr18077.46](https://github.com/hebasto/bitcoin/commits/pr18077.46)) due to the conflicts with #20527.

hebasto commented:
Rebased cc68169e72c5b28fe53466a86eeac9fd214cd0d1 -> 93ff21a929ba3036d6cd5d822f300cf036ff939f ([pr18077.46](https://github.com/hebasto/bitcoin/commits/pr18077.46) -> [pr18077.47](https://github.com/hebasto/bitcoin/commits/pr18077.47)) due to the conflict with https://github.com/bitcoin-core/gui/pull/115.

LarryRuane commented:
In case any of this is helpful... I wanted to try running this, I'm running Ubuntu 18.04 as a Virtualbox guest on a Windows 10 host running at home (Xfinity-Comcast, nothing unusual), I checked out branch `pr18077.47`, ran
```
$ sudo apt install libminiupnpc-dev libnatpmp-dev
(... successful)
$ ./autogen.sh && ./configure --disable-wallet --without-miniupnpc --enable-natpmp-default && make
(...)
checking whether to build with support for UPnP... no
checking whether to build with support for NAT-PMP... yes
checking whether to build with NAT-PMP enabled by default... yes
(...)
$ src/bitcoind
```
Starts up normally, but this prints to `debug.log` every 5 minutes:
```
2020-12-16T19:14:17Z natpmp: The gateway does not support NAT-PMP.
```
and `bitcoin-cli -netstat` shows that I have no incoming peers. I think this is all as expected, but I'm not testing much. I don't know how to change my gateway to support NAT-PMP (I'd appreciate any help). I did change my cable modem router's firewall to allow P2P applications on both IPv4 and IPv6, but that didn't make any difference.

hebasto commented:
@LarryRuane 

Thanks for testing!

> I don't know how to change my gateway to support NAT-PMP (I'd appreciate any help).

Your gateway s/w manual (if available) could help. My TP-Link Archer C6  does not have NAT-PMP support with stock s/w, so I've flashed the OpenWrt on it.

laanwj commented:
Looked over the overall changes, they look reasonable to me.
Will test.

laanwj commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+#endif // #ifdef USE_UPNP
+
+#ifdef USE_NATPMP
+        // Low priority protocol.
```
Why is NATPMP "low priority"? I think NATPMP is the preferred protocol?

hebasto replied:
https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-588484002 by @luke-jr

laanwj replied:
FWIW, I disagree. I think we should prefer NATPMP and phase out UPnP in the longer run when it has proven to be stable.

vasild replied:
@luke-jr, what do you mean by "maps random ports often"? Some implementation of natpmp has bugs? Can you give an example?

Maybe just remove this comment "// Low priority protocol." -- it is unclear what is meant by it and is missing in other places where `#ifdef USE_NATPMP` is used.

hebasto replied:
@vasild 
> @luke-jr, what do you mean by "maps random ports often"? Some implementation of natpmp has bugs? Can you give an example?

Some NAT-PMP daemon implementations do randomize an external port. For example, see https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-588121365 and http://bodgitandscarper.co.uk/natpmpd/:
> The external TCP or UDP port assigned to the client is always randomised rather than giving the first client the port it actually requested and then trying to work out what to do for additional clients that want the same external port.

vasild replied:
@hebasto, Thanks for the clarification! IMO the nat-pmp implementation overriding the desired external port is fine as long as we advertise the overriden one instead of the desired one (I guess we do). That should also be expected as it is allowed by the RFC.

I guess something like that happens with UPnP too if the desired port is already used?

hebasto replied:
> IMO the nat-pmp implementation overriding the desired external port is fine as long as we advertise the overriden one instead of the desired one (I guess we do).

Yes, we do:
https://github.com/bitcoin/bitcoin/blob/890e59c64d62fd39778ede620a70fe6bc60d9993/src/mapport.cpp#L97-L102

> I guess something like that happens with UPnP too if the desired port is already used?

https://github.com/bitcoin/bitcoin/blob/890e59c64d62fd39778ede620a70fe6bc60d9993/src/mapport.cpp#L173-L186

laanwj commented:
It works !

- On Ubuntu 20.04 I installed the `libnatpmp-dev` package, re-ran configure, it was detected correctly and enabled
- I run bitcoind with `-natpmp`

I see the following in my log
```
2020-12-17T15:06:10Z natpmp: Port mapping successful. External address = x.x.x.x:18444
```
My router's admin interface show the redirect:
```
Protocol | External Port | Client Address | Client Port |  
-- | -- | -- | -- | --
TCP | 18444 | 192.168.x.x | 18444
```

laanwj commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+#ifdef USE_NATPMP
+#include <natpmp.h>
+#ifdef WIN32
+#include <winsock2.h>
```
I think we can avoid this platform dependent include dance by including `<src/compat.h>`.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-748361576).

laanwj commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+        } else {
+            LogPrintf("natpmp: readnatpmpresponseorretry() for public address failed with %d error.\n", r_read);
+        }
+
```
Spurious newline?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-748361576).

laanwj commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
```
```suggestion
        if (r != 0) FreeUPNPUrls(&urls);
```

laanwj replied:
Oh, this is a move-only part, ok.

laanwj commented on src/mapport.cpp:
```diff
@@ -0,0 +1,342 @@
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
```
I guess at some point in the future we could replace this thread (which does nothing but waiting and periodically run stuff) with scheduler runnables instead? (not in this PR)

hebasto commented:
Updated 93ff21a929ba3036d6cd5d822f300cf036ff939f -> 890e59c64d62fd39778ede620a70fe6bc60d9993 ([pr18077.47](https://github.com/hebasto/bitcoin/commits/pr18077.47) -> [pr18077.48](https://github.com/hebasto/bitcoin/commits/pr18077.48)):

- rebased due to the conflict with #20681
- addressed the recent @laanwj's comments
- used C++14 chrono literal

hebasto commented:
Rebased 890e59c64d62fd39778ede620a70fe6bc60d9993 -> a191e23b8e7f0e19fc0359825eb7ca0d47966fa9 ([pr18077.48](https://github.com/hebasto/bitcoin/commits/pr18077.48) -> [pr18077.49](https://github.com/hebasto/bitcoin/commits/pr18077.49)) due to the conflict with #20864.

laanwj commented:
Tested and code review ACK a191e23b8e7f0e19fc0359825eb7ca0d47966fa9

ryanofsky commented on src/qt/optionsdialog.cpp:
```diff
@@ -50,6 +51,10 @@ OptionsDialog::OptionsDialog(QWidget *parent, bool enableWallet) :
 #endif
+    connect(this, &QDialog::accepted, [this](){
+        QSettings settings;
+        model->node().mapPort(settings.value("fUseUPnP").toBool());
```
In commit "gui: Apply port mapping changes on dialog exit" (58e8364dcdc4e57b0caac09f8402e6535301de9b)

@hebasto It seems like this commit has a bug and will incorrectly overwrite command line and config file `-upnp` and `-natpmp` options. I think this might have happened before, but previously only if these settings were changed, now if the dialog is just opened and closed.

Would you have an opinion on reverting this change? Aside from the bug, it seems simpler to go back to fully applying settings in OptionsModel, instead of using half-using OptionsModel to apply the setting, and then reading the setting later in a different code path and QSettings instance in dialog code outside of OptionsModel.

(Encountered this while rebasing #15936)

hebasto replied:
@ryanofsky

You are right.

> It seems like this commit has a bug and will incorrectly overwrite command line and config file `-upnp` and `-natpmp` options.

The concerns about user-friendly and correct behavior were raised in #18184. Now users are able to switch `-upnp` and `-natpmp` on/off on-the-fly via the GUI. Should we drop this feature?

> I think this might have happened before, but previously only if these settings were changed, now if the dialog is just opened and closed.

Exiting the dialog via the "Cancel" push button or pressing the "Esc" key should not trigger that code path.

Anyway, I'll happy to revert (or enhance) that change in a proper way.
